### DISCUSSION TITLE
Arguments can't be autocompleted as files on Typer.

### DISCUSSION THREAD

**Discussion by i30817** [fastapi/typer#625] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional, List
from typer import style, echo, Typer, Exit, Argument, Option, Context

def complete_path(ctx: Context, args: List[str], incomplete: str):
    #usually when autocompleting, if you return a empty list,
    #typer will allow the console to autocomplete, except that
    #for arguments, a space is added no matter what you type, which ruins it
    return []

def single(cfg: Path = Argument(None, exists=True, file_okay=True, dir_okay=False, readable=True, resolve_path=True,shell_complete=complete_path)):
    pass

if __name__ == "__main__":
    app = Typer(add_completion=True)
    app.command()(single)
    app()
```

### Description

As the example mentions, autocomplete for paths or files depends on the text cursor not moving. Whenever you try to autocomplete a typer.Argument, except the first time (before you write anything and incomplete is empty), even if you return a empty list you get a space added to the end. If you do that (return a empty list) in a Option shell_complete function, it instead fall-backs to the  cmd line path autocomplete correctly, with no space added which allows autocomplete to function.

I don't get this inconsistency. Actually it might not even be adding the autocomplete to the argument, because i just printed something there and nothing happened.

I _added_ the autocomplete in a attempt to workaround this behavior, which is the default.

edit: shell_complete doesn't work as a argument to ... Argument, but autocompletion=complete_path does. In that case, the program doesn't add a space at the end and it works as expected.

### Operating System

Linux

### Typer Version

0.9.0

### Python Version

Python 3.10.6

**Comment by i30817** [comment]
Maybe it's on purpose that any path is considered 'complete' even paths that don't exist? If so, maybe specifying exists=True could modify that?

**Reply by maszczyn** [reply]
I don't think it's related to `exists` flag. This flag is taken from Click, which Typer uses a lot to handle the CLI, including shell completions, and as far as I checked, Click completes File & Path arguments as expected, even with the default of `exists=False`. I think your issue might be related to what I observed and [started a discussion](https://github.com/tiangolo/typer/discussions/682) for.

**Reply by i30817** [reply]
I'd forgotten this issue. What I ended up doing is use the empty list autocompletition for 'can be anywhere' file Arguments (works if open quotes before trying) and shell_complete for file options I already knew all possibilities for (fixed directory given access to the first argument or its default).

I don't think there is anything _wrong_ specifically, except autocompletition and shell_complete doing different things in spite of being called equivalent in the documentation (I used autocompletion for the file argument and shell_complete for the option arguments). And maybe the default for files completion being buggy and adding a space to any 'completion' instead of trying to complete without that trick (and not working without a opening quote if the path has spaces).

**Reply by svlandeg** [reply]
> And maybe the default for files completion being buggy and adding a space to any 'completion' instead of trying to complete without that trick (and not working without a opening quote if the path has spaces).

This definitely feels like a bug in Typer, I'm trying to hunt down how/where this happens exactly. Related Issue: https://github.com/fastapi/typer/issues/951

**Comment by svlandeg** [comment, marked as answer]
This has been fixed by https://github.com/fastapi/typer/pull/1138, and is available since Typer v0.15.3 🎉
